<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function to do colocalisation tests of two traits — coloc.test • coloc</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Function to do colocalisation tests of two traits — coloc.test" />

<meta property="og:description" content="Performs the colocalisation tests described in Plagnol et al (2009) and
Wallace et al (2012)." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">coloc</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">3.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/colocqq-tests.html">Test for the single-&gt; joint linear regression calculations</a>
    </li>
    <li>
      <a href="../articles/vignette.html">Coloc: a package for colocalisation analyses</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/chr1swallace/coloc">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Function to do colocalisation tests of two traits</h1>
    <small class="dont-index">Source: <a href='https://github.com/chr1swallace/coloc/blob/master/R/coloc.test.R'><code>R/coloc.test.R</code></a></small>
    <div class="hidden name"><code>coloc.test.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Performs the colocalisation tests described in Plagnol et al (2009) and
Wallace et al (2012).</p>
    
    </div>

    <pre class="usage"><span class='fu'>coloc.test</span>(<span class='no'>X</span>, <span class='no'>Y</span>, <span class='kw'>vars.drop</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>X</th>
      <td><p>Either an lm or glm object for trait 1.  The intersection of
<code>names(coefficients(X))</code> and <code>names(coefficients(Y))</code> is used to
identify SNPs in common which will be tested for colocalisation.  Any
Intercept term is dropped, but other covariates should have distinct names or
be listed in <code>vars.drop</code> to avoid them being included in the
colocalisation test.</p></td>
    </tr>
    <tr>
      <th>Y</th>
      <td><p>Either an lm or glm object for trait 2.</p></td>
    </tr>
    <tr>
      <th>vars.drop</th>
      <td><p>Character vector naming additional variables in either
regression which are not SNPs and should not be used in the colocalisation
test.  They should appear in
<code>c(names(coefficients(X)),names(coefficients(Y)))</code></p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>other arguments passed to <code><a href='coloc.test.summary.html'>coloc.test.summary</a>()</code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a numeric vector with 3 named elements:</p>
<dt>eta.hat</dt><dd><p>The estimated slope.</p></dd>
<dt>chisquare</dt><dd><p>The chisquared test statistic</p></dd>
<dt>n</dt><dd><p>The number of snps used in the test.  If eta were known, this
would be the degrees of freedom of the test. Because eta has been replaced by
its ML estimate, Plagnol et al suggest we expect the degrees of freedom to be
n-1, but this requires the likelihood to be well behaved which is not always
the case.  We prefer to consider the posterior predictive p value.</p></dd>
<dt>ppp</dt><dd><p>The posterior predictive p value</p></dd>

    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This is a test for proportionality of regression coefficients from two
independent regressions.  Analysis can either be based on a profile
likelihood approach, where the proportionality coefficient, <code>eta</code>, is
replaced by its maximum likelihood value, and inference is based on a
chisquare test (<code>p.value</code>), or taking a hybrid-Bayesian approach and
integrating the p value over the posterior distribution of <code>eta</code>, which
gives a posterior predictive p value.  The Bayesian approach can also be used
to give a credible interval for <code>eta</code>.  See the references below for
further details.</p>
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>Plagnol et al's original test was available in his R package
<code>QTLMatch v0.8</code> which now appears unavailable.  The numerically
identical test, extended to allow for more than two SNPs, can be found in
this package by looking at the chisquare statistic and the degrees of freedom
given by <code>chisquare()</code> and <code>df()</code> respectively.</p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Wallace et al (2012).  Statistical colocalisation of monocyte
gene expression and genetic risk variants for type 1 diabetes.  Hum Mol Genet
21:2815-2824.  <a href='http://europepmc.org/abstract/MED/22403184'>http://europepmc.org/abstract/MED/22403184</a></p>
<p>Plagnol et al (2009).  Statistical independence of the colocalized
association signals for type 1 diabetes and RPS26 gene expression on
chromosome 12q13. Biostatistics 10:327-34.
<a href='http://www.ncbi.nlm.nih.gov/pubmed/19039033'>http://www.ncbi.nlm.nih.gov/pubmed/19039033</a></p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
  <span class='co'>## simulate covariate matrix (X) and continuous response vector (Y)</span>
  <span class='co'>## for two populations/triats Y1 and Y2 depend equally on f1 and f2</span>
  <span class='co'>## within each population, although their distributions differ between</span>
  <span class='co'>## populations.  They are compatible with a null hypothesis that they</span>
  <span class='co'>## share a common causal variant</span>
<span class='fu'>set.seed</span>(<span class='fl'>1</span>)
  <span class='no'>X1</span> <span class='kw'>&lt;-</span> <span class='fu'>matrix</span>(<span class='fu'>rbinom</span>(<span class='fl'>1000</span>,<span class='fl'>1</span>,<span class='fl'>0.4</span>),<span class='kw'>ncol</span><span class='kw'>=</span><span class='fl'>2</span>)
  <span class='no'>Y1</span> <span class='kw'>&lt;-</span> <span class='fu'>rnorm</span>(<span class='fl'>500</span>,<span class='fu'>apply</span>(<span class='no'>X1</span>,<span class='fl'>1</span>,<span class='no'>sum</span>),<span class='fl'>2</span>)
  <span class='no'>X2</span> <span class='kw'>&lt;-</span> <span class='fu'>matrix</span>(<span class='fu'>rbinom</span>(<span class='fl'>1000</span>,<span class='fl'>1</span>,<span class='fl'>0.6</span>),<span class='kw'>ncol</span><span class='kw'>=</span><span class='fl'>2</span>)
  <span class='no'>Y2</span> <span class='kw'>&lt;-</span> <span class='fu'>rnorm</span>(<span class='fl'>500</span>,<span class='fl'>2</span>*<span class='fu'>apply</span>(<span class='no'>X2</span>,<span class='fl'>1</span>,<span class='no'>sum</span>),<span class='fl'>5</span>)

  <span class='fu'>boxplot</span>(<span class='fu'>list</span>(<span class='no'>Y1</span>,<span class='no'>Y2</span>),<span class='kw'>names</span><span class='kw'>=</span><span class='fu'>c</span>(<span class='st'>"Y1"</span>,<span class='st'>"Y2"</span>))</div><div class='img'><img src='coloc.test-1.png' alt='' width='700' height='433' /></div><div class='input'>
  <span class='co'>## fit and store linear model objects</span>
  <span class='fu'>colnames</span>(<span class='no'>X1</span>) <span class='kw'>&lt;-</span> <span class='fu'>colnames</span>(<span class='no'>X2</span>) <span class='kw'>&lt;-</span> <span class='fu'>c</span>(<span class='st'>"f1"</span>,<span class='st'>"f2"</span>)
  <span class='fu'><a href='eta.html'>summary</a></span>(<span class='no'>lm1</span> <span class='kw'>&lt;-</span> <span class='fu'>lm</span>(<span class='no'>Y1</span>~<span class='no'>f1</span>+<span class='no'>f2</span>,<span class='kw'>data</span><span class='kw'>=</span><span class='fu'>as.data.frame</span>(<span class='no'>X1</span>)))</div><div class='output co'>#&gt; 
#&gt; Call:
#&gt; lm(formula = Y1 ~ f1 + f2, data = as.data.frame(X1))
#&gt; 
#&gt; Residuals:
#&gt;     Min      1Q  Median      3Q     Max 
#&gt; -5.8008 -1.5027  0.0237  1.5084  6.3045 
#&gt; 
#&gt; Coefficients:
#&gt;             Estimate Std. Error t value Pr(&gt;|t|)    
#&gt; (Intercept)  -0.1931     0.1409  -1.370    0.171    
#&gt; f1            1.0331     0.1959   5.275 1.99e-07 ***
#&gt; f2            1.2153     0.1928   6.305 6.38e-10 ***
#&gt; ---
#&gt; Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
#&gt; 
#&gt; Residual standard error: 2.116 on 497 degrees of freedom
#&gt; Multiple R-squared:  0.1276,	Adjusted R-squared:  0.1241 
#&gt; F-statistic: 36.36 on 2 and 497 DF,  p-value: 1.838e-15
#&gt; </div><div class='input'>  <span class='fu'><a href='eta.html'>summary</a></span>(<span class='no'>lm2</span> <span class='kw'>&lt;-</span> <span class='fu'>lm</span>(<span class='no'>Y2</span>~<span class='no'>f1</span>+<span class='no'>f2</span>,<span class='kw'>data</span><span class='kw'>=</span><span class='fu'>as.data.frame</span>(<span class='no'>X2</span>)))</div><div class='output co'>#&gt; 
#&gt; Call:
#&gt; lm(formula = Y2 ~ f1 + f2, data = as.data.frame(X2))
#&gt; 
#&gt; Residuals:
#&gt;      Min       1Q   Median       3Q      Max 
#&gt; -14.7797  -3.2392   0.0128   3.7783  15.9367 
#&gt; 
#&gt; Coefficients:
#&gt;             Estimate Std. Error t value Pr(&gt;|t|)    
#&gt; (Intercept)  -0.1150     0.5157  -0.223 0.823674    
#&gt; f1            2.0334     0.4965   4.095 4.92e-05 ***
#&gt; f2            1.9147     0.4976   3.848 0.000135 ***
#&gt; ---
#&gt; Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
#&gt; 
#&gt; Residual standard error: 5.362 on 497 degrees of freedom
#&gt; Multiple R-squared:  0.05575,	Adjusted R-squared:  0.05195 
#&gt; F-statistic: 14.67 on 2 and 497 DF,  p-value: 6.442e-07
#&gt; </div><div class='input'>
  <span class='co'>## test whether the traits are compatible with colocalisation</span>
  <span class='co'>### ppp should be large (&gt;0.05, for example), indicating that they are.</span>
  <span class='fu'>par</span>(<span class='kw'>mfrow</span><span class='kw'>=</span><span class='fu'>c</span>(<span class='fl'>2</span>,<span class='fl'>2</span>))
  <span class='no'>obj</span> <span class='kw'>&lt;-</span> <span class='fu'>coloc.test</span>(<span class='no'>lm1</span>,<span class='no'>lm2</span>,
             <span class='kw'>plots.extra</span><span class='kw'>=</span><span class='fu'>list</span>(<span class='kw'>x</span><span class='kw'>=</span><span class='fu'>c</span>(<span class='st'>"eta"</span>,<span class='st'>"theta"</span>),
                              <span class='kw'>y</span><span class='kw'>=</span><span class='fu'>c</span>(<span class='st'>"lhood"</span>,<span class='st'>"lhood"</span>)))
  <span class='fu'><a href='plot-methods.html'>plot</a></span>(<span class='no'>obj</span>)</div><div class='img'><img src='coloc.test-2.png' alt='' width='700' height='433' /></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#note">Note</a></li>

      <li><a href="#references">References</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Chris Wallace

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Chris Wallace, Claudia Giambartolomei.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

