<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Wrapper to use colocalization testing within a Bayesian model averaging
structure. — coloc.bma • coloc</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Wrapper to use colocalization testing within a Bayesian model averaging
structure. — coloc.bma" />

<meta property="og:description" content="Performs the colocalisation tests described in Plagnol et al (2009) and
Wallace et al (2012)." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">coloc</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">3.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/colocqq-tests.html">Test for the single-&gt; joint linear regression calculations</a>
    </li>
    <li>
      <a href="../articles/vignette.html">Coloc: a package for colocalisation analyses</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/chr1swallace/coloc">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Wrapper to use colocalization testing within a Bayesian model averaging
structure.</h1>
    <small class="dont-index">Source: <a href='https://github.com/chr1swallace/coloc/blob/master/R/bma.R'><code>R/bma.R</code></a></small>
    <div class="hidden name"><code>coloc.bma.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Performs the colocalisation tests described in Plagnol et al (2009) and
Wallace et al (2012).</p>
    
    </div>

    <pre class="usage"><span class='fu'>coloc.bma</span>(<span class='no'>df1</span>, <span class='no'>df2</span>, <span class='kw'>snps</span> <span class='kw'>=</span> <span class='fu'>intersect</span>(<span class='fu'>setdiff</span>(<span class='fu'>colnames</span>(<span class='no'>df1</span>), <span class='fu'>c</span>(<span class='no'>response1</span>,
  <span class='no'>stratum1</span>)), <span class='fu'>setdiff</span>(<span class='fu'>colnames</span>(<span class='no'>df2</span>), <span class='fu'>c</span>(<span class='no'>response2</span>, <span class='no'>stratum2</span>))),
  <span class='kw'>response1</span> <span class='kw'>=</span> <span class='st'>"Y"</span>, <span class='kw'>response2</span> <span class='kw'>=</span> <span class='st'>"Y"</span>, <span class='kw'>stratum1</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>stratum2</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>family1</span> <span class='kw'>=</span> <span class='st'>"binomial"</span>, <span class='kw'>family2</span> <span class='kw'>=</span> <span class='st'>"binomial"</span>,
  <span class='kw'>bayes</span> <span class='kw'>=</span> !<span class='fu'>is.null</span>(<span class='no'>bayes.factor</span>), <span class='kw'>thr</span> <span class='kw'>=</span> <span class='fl'>0.01</span>, <span class='kw'>nsnps</span> <span class='kw'>=</span> <span class='fl'>2</span>,
  <span class='kw'>n.approx</span> <span class='kw'>=</span> <span class='fl'>1001</span>, <span class='kw'>bayes.factor</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>plot.coeff</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>r2.trim</span> <span class='kw'>=</span> <span class='fl'>0.95</span>, <span class='kw'>quiet</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>bma</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>df1, df2</th>
      <td><p>Each is a dataframe, containing response and
potential explanatory variables for two independent datasets.</p></td>
    </tr>
    <tr>
      <th>snps</th>
      <td><p>The SNPs to consider as potential explanatory variables</p></td>
    </tr>
    <tr>
      <th>response1, response2</th>
      <td><p>The names of the response variables in
<code>df1</code> and <code>df2</code> respectively</p></td>
    </tr>
    <tr>
      <th>stratum1</th>
      <td><p>optional column name of df1 that gives stratum information</p></td>
    </tr>
    <tr>
      <th>stratum2</th>
      <td><p>optional column name of df2 that gives stratum information</p></td>
    </tr>
    <tr>
      <th>family1, family2</th>
      <td><p>the error family for use in <code>glm</code></p></td>
    </tr>
    <tr>
      <th>bayes</th>
      <td><p>Logical, indicating whether to perform Bayesian
inference for the coefficient of proportionality, eta.  If
<code>bayes.factor</code> is supplied, Bayes factors are additionally
computed for the specificed values.  This can add a little time as
it requires numerical integration, so can be set to FALSE to save
time in simulations, for example.</p></td>
    </tr>
    <tr>
      <th>thr</th>
      <td><p>posterior probability threshold used to trim SNP list.
Only SNPs with a marginal posterior probability of inclusion
greater than this with one or other trait will be included in
the full BMA analysis</p></td>
    </tr>
    <tr>
      <th>nsnps</th>
      <td><p>number of SNPs required to model both traits.  The BMA
analysis will average over all possible <code>nsnp</code> SNP models,
subject to <code>thr</code> above.</p></td>
    </tr>
    <tr>
      <th>n.approx</th>
      <td><p>number of values at which to numerically
approximate the posterior</p></td>
    </tr>
    <tr>
      <th>bayes.factor</th>
      <td><p>if true, compare specific models</p></td>
    </tr>
    <tr>
      <th>plot.coeff</th>
      <td><p>deprecated</p></td>
    </tr>
    <tr>
      <th>r2.trim</th>
      <td><p>for pairs SNPs with r2&gt;<code>r2.trim</code>, only one SNP
will be retained.  This avoids numerical instability problems
caused by including two highly correlated SNPs in the model.</p></td>
    </tr>
    <tr>
      <th>quiet</th>
      <td><p>suppress messages about how the model spaced is
trimmed for BMA</p></td>
    </tr>
    <tr>
      <th>bma</th>
      <td><p>if true (default), average over models</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>other parameters passed to <code>coloc.test</code></p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a <code>coloc</code> or <code>colocBayes</code> object</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This is a test for proportionality of regression coefficients from two
independent regressions.  Analysis can either be based on a profile
likelihood approach, where the proportionality coefficient, <code>eta</code>, is
replaced by its maximum likelihood value, and inference is based on a
chisquare test (<code>p.value</code>), or taking a hybrid-Bayesian approach and
integrating the p value over the posterior distribution of <code>eta</code>, which
gives a posterior predictive p value.  The Bayesian approach can also be used
to give a credible interval for <code>eta</code>.  See the references below for
further details.</p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Wallace et al (2012).  Statistical colocalisation of
    monocyte gene expression and genetic risk variants for type 1
    diabetes.  Hum Mol Genet 21:2815-2824.
    <a href='http://europepmc.org/abstract/MED/22403184'>http://europepmc.org/abstract/MED/22403184</a></p>
<p>Plagnol et al (2009).  Statistical independence of the colocalized
association signals for type 1 diabetes and RPS26 gene expression on
chromosome 12q13. Biostatistics 10:327-34.
<a href='http://www.ncbi.nlm.nih.gov/pubmed/19039033'>http://www.ncbi.nlm.nih.gov/pubmed/19039033</a></p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
 <span class='co'>## simulate covariate matrix (X) and continuous response vector (Y)</span>
 <span class='co'>## for two populations/triats Y1 and Y2 depend equally on f1 and f2</span>
 <span class='co'>## within each population, although their distributions differ between</span>
 <span class='co'>## populations.  They are compatible with a null hypothesis that they</span>
 <span class='co'>## share a common causal variant</span>
<span class='fu'>set.seed</span>(<span class='fl'>1</span>)
 <span class='no'>X1</span> <span class='kw'>&lt;-</span> <span class='fu'>matrix</span>(<span class='fu'>rbinom</span>(<span class='fl'>2000</span>,<span class='fl'>1</span>,<span class='fl'>0.4</span>),<span class='kw'>ncol</span><span class='kw'>=</span><span class='fl'>4</span>)
 <span class='no'>Y1</span> <span class='kw'>&lt;-</span> <span class='fu'>rnorm</span>(<span class='fl'>500</span>,<span class='fu'>rowSums</span>(<span class='no'>X1</span>[,<span class='fl'>1</span>:<span class='fl'>2</span>]),<span class='fl'>2</span>)
 <span class='no'>X2</span> <span class='kw'>&lt;-</span> <span class='fu'>matrix</span>(<span class='fu'>rbinom</span>(<span class='fl'>2000</span>,<span class='fl'>1</span>,<span class='fl'>0.6</span>),<span class='kw'>ncol</span><span class='kw'>=</span><span class='fl'>4</span>)
 <span class='no'>Y2</span> <span class='kw'>&lt;-</span> <span class='fu'>rnorm</span>(<span class='fl'>500</span>,<span class='fu'>rowSums</span>(<span class='no'>X2</span>[,<span class='fl'>1</span>:<span class='fl'>2</span>]),<span class='fl'>5</span>)

 <span class='fu'>boxplot</span>(<span class='fu'>list</span>(<span class='no'>Y1</span>,<span class='no'>Y2</span>),<span class='kw'>names</span><span class='kw'>=</span><span class='fu'>c</span>(<span class='st'>"Y1"</span>,<span class='st'>"Y2"</span>))</div><div class='img'><img src='coloc.bma-1.png' alt='' width='700' height='433' /></div><div class='input'>
 <span class='co'>## fit and store linear model objects</span>
 <span class='fu'>colnames</span>(<span class='no'>X1</span>) <span class='kw'>&lt;-</span> <span class='fu'>colnames</span>(<span class='no'>X2</span>) <span class='kw'>&lt;-</span> <span class='fu'>sprintf</span>(<span class='st'>"f%s"</span>,<span class='fl'>1</span>:<span class='fu'>ncol</span>(<span class='no'>X1</span>))
 <span class='fu'><a href='eta.html'>summary</a></span>(<span class='no'>lm1</span> <span class='kw'>&lt;-</span> <span class='fu'>lm</span>(<span class='no'>Y1</span>~<span class='no'>f1</span>+<span class='no'>f2</span>+<span class='no'>f3</span>+<span class='no'>f4</span>,<span class='kw'>data</span><span class='kw'>=</span><span class='fu'>as.data.frame</span>(<span class='no'>X1</span>)))</div><div class='output co'>#&gt; 
#&gt; Call:
#&gt; lm(formula = Y1 ~ f1 + f2 + f3 + f4, data = as.data.frame(X1))
#&gt; 
#&gt; Residuals:
#&gt;     Min      1Q  Median      3Q     Max 
#&gt; -6.4771 -1.4158 -0.0845  1.3748  7.1976 
#&gt; 
#&gt; Coefficients:
#&gt;             Estimate Std. Error t value Pr(&gt;|t|)    
#&gt; (Intercept)   0.2863     0.1680   1.704  0.08903 .  
#&gt; f1            0.6611     0.1855   3.565  0.00040 ***
#&gt; f2            0.5185     0.1819   2.850  0.00455 ** 
#&gt; f3            0.3080     0.1851   1.664  0.09670 .  
#&gt; f4           -0.2047     0.1823  -1.123  0.26196    
#&gt; ---
#&gt; Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
#&gt; 
#&gt; Residual standard error: 1.996 on 495 degrees of freedom
#&gt; Multiple R-squared:  0.04906,	Adjusted R-squared:  0.04137 
#&gt; F-statistic: 6.384 on 4 and 495 DF,  p-value: 5.152e-05
#&gt; </div><div class='input'> <span class='fu'><a href='eta.html'>summary</a></span>(<span class='no'>lm2</span> <span class='kw'>&lt;-</span> <span class='fu'>lm</span>(<span class='no'>Y2</span>~<span class='no'>f1</span>+<span class='no'>f2</span>+<span class='no'>f3</span>+<span class='no'>f4</span>,<span class='kw'>data</span><span class='kw'>=</span><span class='fu'>as.data.frame</span>(<span class='no'>X2</span>)))</div><div class='output co'>#&gt; 
#&gt; Call:
#&gt; lm(formula = Y2 ~ f1 + f2 + f3 + f4, data = as.data.frame(X2))
#&gt; 
#&gt; Residuals:
#&gt;      Min       1Q   Median       3Q      Max 
#&gt; -13.6564  -3.1427   0.0631   3.2446  13.5783 
#&gt; 
#&gt; Coefficients:
#&gt;             Estimate Std. Error t value Pr(&gt;|t|)  
#&gt; (Intercept)  -0.8462     0.6420  -1.318   0.1881  
#&gt; f1            1.0047     0.4658   2.157   0.0315 *
#&gt; f2            1.0072     0.4736   2.127   0.0339 *
#&gt; f3            0.5396     0.4735   1.139   0.2551  
#&gt; f4            1.0390     0.4663   2.228   0.0263 *
#&gt; ---
#&gt; Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
#&gt; 
#&gt; Residual standard error: 5.119 on 495 degrees of freedom
#&gt; Multiple R-squared:  0.02742,	Adjusted R-squared:  0.01957 
#&gt; F-statistic: 3.489 on 4 and 495 DF,  p-value: 0.007988
#&gt; </div><div class='input'>

<span class='co'>## test colocalisation using bma</span>
<span class='no'>df1</span><span class='kw'>=</span><span class='fu'>as.data.frame</span>(<span class='fu'>cbind</span>(<span class='kw'>Y1</span><span class='kw'>=</span><span class='no'>Y1</span>,<span class='no'>X1</span>))
<span class='no'>df2</span><span class='kw'>=</span><span class='fu'>as.data.frame</span>(<span class='fu'>cbind</span>(<span class='kw'>Y2</span><span class='kw'>=</span><span class='no'>Y2</span>,<span class='no'>X2</span>))

<span class='no'>result</span> <span class='kw'>&lt;-</span> <span class='fu'>coloc.bma</span>( <span class='no'>df1</span>, <span class='no'>df2</span>, <span class='kw'>snps</span><span class='kw'>=</span><span class='fu'>colnames</span>(<span class='no'>X1</span>), <span class='kw'>response1</span><span class='kw'>=</span><span class='st'>"Y1"</span>, <span class='kw'>response2</span><span class='kw'>=</span><span class='st'>"Y2"</span>,
<span class='kw'>family1</span><span class='kw'>=</span><span class='st'>"gaussian"</span>, <span class='kw'>family2</span><span class='kw'>=</span><span class='st'>"gaussian"</span>,
<span class='kw'>nsnps</span><span class='kw'>=</span><span class='fl'>2</span>,<span class='kw'>bayes.factor</span><span class='kw'>=</span><span class='fu'>c</span>(<span class='fl'>1</span>,<span class='fl'>2</span>,<span class='fl'>3</span>))</div><div class='output co'>#&gt; <span class='message'>Dropped 0 of 4 SNPs due to LD: r2 &gt; 0.95.</span></div><div class='output co'>#&gt; <span class='message'>4 SNPs remain.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='message'>Restricting model space.</span></div><div class='output co'>#&gt; <span class='message'>2 SNPs have single SNP posterior probabilities &lt; 0.01</span></div><div class='output co'>#&gt; <span class='message'>Models containing only these SNPs will not be explored.</span></div><div class='output co'>#&gt; Fitting 5 multi SNP models to dataset 1 </div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; Fitting 5 multi SNP models to dataset 2 </div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; Averaging coloc testing over 3 models with posterior probabilities &gt;= 2e-08 
#&gt; ...</div><div class='input'><span class='no'>result</span></div><div class='output co'>#&gt;   eta.hat         n ppp.value 
#&gt; 1.6564113 2.0000000 0.8085261 </div><div class='input'><span class='fu'><a href='plot-methods.html'>plot</a></span>(<span class='no'>result</span>)</div><div class='img'><img src='coloc.bma-2.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'>## test colocalisation when one dataset contains a stratifying factor in column named "s"</span>
<span class='no'>df1</span>$<span class='no'>s</span> <span class='kw'>&lt;-</span> <span class='fu'>rbinom</span>(<span class='fl'>500</span>,<span class='fl'>1</span>,<span class='fl'>0.5</span>)
<span class='no'>result</span> <span class='kw'>&lt;-</span> <span class='fu'>coloc.bma</span>( <span class='no'>df1</span>, <span class='no'>df2</span>, <span class='kw'>snps</span><span class='kw'>=</span><span class='fu'>colnames</span>(<span class='no'>X1</span>), <span class='kw'>response1</span><span class='kw'>=</span><span class='st'>"Y1"</span>, <span class='kw'>response2</span><span class='kw'>=</span><span class='st'>"Y2"</span>,
<span class='kw'>stratum1</span><span class='kw'>=</span><span class='st'>"s"</span>,
<span class='kw'>family1</span><span class='kw'>=</span><span class='st'>"gaussian"</span>, <span class='kw'>family2</span><span class='kw'>=</span><span class='st'>"gaussian"</span>,
<span class='kw'>nsnps</span><span class='kw'>=</span><span class='fl'>2</span>,<span class='kw'>bayes.factor</span><span class='kw'>=</span><span class='fu'>c</span>(<span class='fl'>1</span>,<span class='fl'>2</span>,<span class='fl'>3</span>))</div><div class='output co'>#&gt; <span class='message'>Dropped 0 of 4 SNPs due to LD: r2 &gt; 0.95.</span></div><div class='output co'>#&gt; <span class='message'>4 SNPs remain.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='message'>Restricting model space.</span></div><div class='output co'>#&gt; <span class='message'>1 SNPs have single SNP posterior probabilities &lt; 0.01</span></div><div class='output co'>#&gt; <span class='message'>Models containing only these SNPs will not be explored.</span></div><div class='output co'>#&gt; Fitting 6 multi SNP models to dataset 1 </div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; Fitting 6 multi SNP models to dataset 2 </div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Recycling array of length 1 in vector-array arithmetic is deprecated.</span>
#&gt; <span class='warning'>  Use c() or as.vector() instead.</span></div><div class='output co'>#&gt; Averaging coloc testing over 3 models with posterior probabilities &gt;= 2.2e-08 
#&gt; ...</div><div class='input'><span class='no'>result</span></div><div class='output co'>#&gt;   eta.hat         n ppp.value 
#&gt;  1.658547  2.000000  0.810177 </div><div class='input'><span class='fu'><a href='plot-methods.html'>plot</a></span>(<span class='no'>result</span>)</div><div class='img'><img src='coloc.bma-3.png' alt='' width='700' height='433' /></div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#references">References</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Chris Wallace

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Chris Wallace, Claudia Giambartolomei.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

