<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Bayesian finemapping analysis — finemap.abf • coloc</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Bayesian finemapping analysis — finemap.abf" />

<meta property="og:description" content="Bayesian finemapping analysis" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">coloc</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">3.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/colocqq-tests.html">Test for the single-&gt; joint linear regression calculations</a>
    </li>
    <li>
      <a href="../articles/vignette.html">Coloc: a package for colocalisation analyses</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/chr1swallace/coloc">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Bayesian finemapping analysis</h1>
    <small class="dont-index">Source: <a href='https://github.com/chr1swallace/coloc/blob/master/R/claudia.R'><code>R/claudia.R</code></a></small>
    <div class="hidden name"><code>finemap.abf.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Bayesian finemapping analysis</p>
    
    </div>

    <pre class="usage"><span class='fu'>finemap.abf</span>(<span class='no'>dataset</span>, <span class='kw'>p1</span> <span class='kw'>=</span> <span class='fl'>1e-04</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>dataset</th>
      <td><p>a list with the following elements</p><dl class='dl-horizontal'>

  <dt>pvalues</dt><dd><p>P-values for each SNP in dataset 1</p></dd>

  <dt>N</dt><dd><p>Number of samples in dataset 1</p></dd>

  <dt>MAF</dt><dd><p>minor allele frequency of the variants</p></dd>

<dt>beta</dt><dd><p>regression coefficient for each SNP from dataset 1</p></dd>

<dt>varbeta</dt><dd><p>variance of beta</p></dd>

<dt>type</dt><dd><p>the type of data in dataset 1 - either "quant" or "cc" to denote quantitative or case-control</p></dd>

<dt>s</dt><dd><p>for a case control dataset, the proportion of samples in dataset 1 that are cases</p></dd>

 <dt>sdY</dt><dd><p>for a quantitative trait, the population standard deviation of the trait.  if not given, it can be estimated from the vectors of varbeta and MAF</p></dd>

<dt>snp</dt><dd><p>a character vector of snp ids, optional. If present, it will be used to merge dataset1 and dataset2.  Otherwise, the function assumes dataset1 and dataset2 contain results for the same SNPs in the same order.</p></dd>

</dl>

<p>Some of these items may be missing, but you must give</p><ul>
<li><p>always<code>type</code></p></li>
<li><p>if <code>type</code>=="cc"<code>s</code></p></li>
<li><p>if <code>type</code>=="quant" and <code>sdY</code> known<code>sdY</code></p></li>
<li><p>if <code>type</code>=="quant" and <code>sdY</code> unknown<code>beta</code>, <code>varbeta</code>, <code>N</code>, <code>MAF</code>
and then either</p></li>
<li><p><code>pvalues</code>, <code>MAF</code></p></li>
<li><p><code>beta</code>, <code>varbeta</code></p></li>
</ul></td>
    </tr>
    <tr>
      <th>p1</th>
      <td><p>prior probability a SNP is associated with the trait 1, default 1e-4</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a <code>data.frame</code>:</p><ul>
<li><p>an annotated version of the input data containing log Approximate Bayes Factors and intermediate calculations, and the posterior probability of the SNP being causal</p></li>
</ul>

    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function calculates posterior probabilities of different
causal variant for a single trait.</p>
<p>If regression coefficients and variances are available, it
calculates Bayes factors for association at each SNP.  If only p
values are available, it uses an approximation that depends on the
SNP's MAF and ignores any uncertainty in imputation.  Regression
coefficients should be used if available.</p>
    

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>
          </ul>

    <h2>Author</h2>
    
Chris Wallace

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Chris Wallace, Claudia Giambartolomei.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

