<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Functions to prepare principle component models for colocalisation testing — pcs.prepare • coloc</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Functions to prepare principle component models for colocalisation testing — pcs.prepare" />

<meta property="og:description" content="Prepares principal components of two datasets for
colocalisation testing." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">coloc</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">3.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/colocqq-tests.html">Test for the single-&gt; joint linear regression calculations</a>
    </li>
    <li>
      <a href="../articles/vignette.html">Coloc: a package for colocalisation analyses</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/chr1swallace/coloc">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Functions to prepare principle component models for colocalisation testing</h1>
    <small class="dont-index">Source: <a href='https://github.com/chr1swallace/coloc/blob/master/R/pcs.R'><code>R/pcs.R</code></a></small>
    <div class="hidden name"><code>pcs.prepare.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Prepares principal components of two datasets for
colocalisation testing.</p>
    
    </div>

    <pre class="usage"><span class='fu'>pcs.prepare</span>(<span class='no'>X1</span>, <span class='no'>X2</span>, <span class='kw'>impute</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>X1</th>
      <td><p>Either a SnpMatrix or numeric matrix of genetic data.
Columns index SNPs, rows index samples.</p></td>
    </tr>
    <tr>
      <th>X2</th>
      <td><p>as X1</p></td>
    </tr>
    <tr>
      <th>impute</th>
      <td><p>if TRUE (default), impute missing genotypes</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a <code>colocPCs</code> object.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>If <code>X1</code> and <code>X2</code> are <code>SnpMatrix</code> objects, they are checked for
missing data, and any missing values imputed by repeated use of
<code>impute.snps</code> from the <code>snpStats</code> package.</p>
<p>Columns with common names are <code>rbind</code>ed together and principal
components calculated using <code>prcomp</code>.</p>
<p><code>pcs.model</code> can then be invoked to create <code>glm</code> objects.</p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Wallace et al (2012).  Statistical colocalisation of monocyte
gene expression and genetic risk variants for type 1 diabetes.  Hum Mol Genet
21:2815-2824.  <a href='http://europepmc.org/abstract/MED/22403184'>http://europepmc.org/abstract/MED/22403184</a></p>
<p>Plagnol et al (2009).  Statistical independence of the colocalized
association signals for type 1 diabetes and RPS26 gene expression on
chromosome 12q13. Biostatistics 10:327-34.
<a href='http://www.ncbi.nlm.nih.gov/pubmed/19039033'>http://www.ncbi.nlm.nih.gov/pubmed/19039033</a></p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
 <span class='co'>## simulate covariate matrix (X) and continuous response vector (Y)</span>
 <span class='co'>## for two populations/triats Y1 and Y2 depend equally on f1 and f2</span>
 <span class='co'>## within each population, although their distributions differ between</span>
 <span class='co'>## populations.  They are compatible with a null hypothesis that they</span>
 <span class='co'>## share a common causal variant, with the effect twice as strong for</span>
 <span class='co'>## Y2 as Y1</span>
<span class='fu'>set.seed</span>(<span class='fl'>1</span>)
 <span class='no'>X1</span> <span class='kw'>&lt;-</span> <span class='fu'>matrix</span>(<span class='fu'>rbinom</span>(<span class='fl'>5000</span>,<span class='fl'>1</span>,<span class='fl'>0.4</span>),<span class='kw'>ncol</span><span class='kw'>=</span><span class='fl'>10</span>)
 <span class='no'>Y1</span> <span class='kw'>&lt;-</span> <span class='fu'>rnorm</span>(<span class='fl'>500</span>,<span class='fu'>apply</span>(<span class='no'>X1</span>[,<span class='fl'>1</span>:<span class='fl'>2</span>],<span class='fl'>1</span>,<span class='no'>sum</span>),<span class='fl'>2</span>)
 <span class='no'>X2</span> <span class='kw'>&lt;-</span> <span class='fu'>matrix</span>(<span class='fu'>rbinom</span>(<span class='fl'>5000</span>,<span class='fl'>1</span>,<span class='fl'>0.6</span>),<span class='kw'>ncol</span><span class='kw'>=</span><span class='fl'>10</span>)
 <span class='no'>Y2</span> <span class='kw'>&lt;-</span> <span class='fu'>rnorm</span>(<span class='fl'>500</span>,<span class='fl'>2</span>*<span class='fu'>apply</span>(<span class='no'>X2</span>[,<span class='fl'>1</span>:<span class='fl'>2</span>],<span class='fl'>1</span>,<span class='no'>sum</span>),<span class='fl'>5</span>)

 <span class='co'>## generate principal components object</span>
 <span class='fu'>colnames</span>(<span class='no'>X1</span>) <span class='kw'>&lt;-</span> <span class='fu'>colnames</span>(<span class='no'>X2</span>) <span class='kw'>&lt;-</span> <span class='fu'>make.names</span>(<span class='fl'>1</span>:<span class='fu'>ncol</span>(<span class='no'>X1</span>))
 <span class='no'>pcs</span> <span class='kw'>&lt;-</span> <span class='fu'>pcs.prepare</span>(<span class='no'>X1</span>,<span class='no'>X2</span>)

 <span class='co'>## generate glm objects</span>
 <span class='no'>m1</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='pcs.model.html'>pcs.model</a></span>(<span class='no'>pcs</span>, <span class='kw'>group</span><span class='kw'>=</span><span class='fl'>1</span>, <span class='kw'>Y</span><span class='kw'>=</span><span class='no'>Y1</span>)</div><div class='output co'>#&gt; selecting 8 components out of 10 to capture 0.8299584 of total variance.</div><div class='input'> <span class='no'>m2</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='pcs.model.html'>pcs.model</a></span>(<span class='no'>pcs</span>, <span class='kw'>group</span><span class='kw'>=</span><span class='fl'>2</span>, <span class='kw'>Y</span><span class='kw'>=</span><span class='no'>Y2</span>)</div><div class='output co'>#&gt; selecting 8 components out of 10 to capture 0.8299584 of total variance.</div><div class='input'>
 <span class='co'>## test colocalisation using PCs</span>
 <span class='fu'><a href='coloc.test.html'>coloc.test</a></span>(<span class='no'>m1</span>,<span class='no'>m2</span>,<span class='kw'>plot.coeff</span><span class='kw'>=</span><span class='fl'>FALSE</span>,<span class='kw'>bayes</span><span class='kw'>=</span><span class='fl'>FALSE</span>)</div><div class='output co'>#&gt;           eta.hat         chisquare                 n p.value.chisquare 
#&gt;         2.8822102         8.7378708         8.0000000         0.2720277 </div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#references">References</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Chris Wallace

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Chris Wallace, Claudia Giambartolomei.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

